File: src/main/java/tlint/InspectionUtil.java
package tlint;

import com.intellij.codeInsight.daemon.impl.HighlightInfoType;
import com.intellij.codeInsight.daemon.impl.SeverityRegistrar;
import com.intellij.lang.annotation.Annotation;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.openapi.editor.colors.EditorColorsManager;
import com.intellij.openapi.editor.colors.EditorColorsScheme;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.openapi.util.TextRange;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class InspectionUtil {
    private InspectionUtil() {
    }

    @NotNull
    private static EditorColorsScheme getColorsScheme(@Nullable EditorColorsScheme customScheme) {
        return customScheme == null ? EditorColorsManager.getInstance().getGlobalScheme() : customScheme;
    }

    @NotNull
    public static TextAttributes getTextAttributes(@Nullable EditorColorsScheme editorColorsScheme, @NotNull SeverityRegistrar severityRegistrar, @NotNull HighlightSeverity severity) {
        TextAttributes textAttributes = severityRegistrar.getTextAttributesBySeverity(severity);
        if (textAttributes == null) {
            EditorColorsScheme colorsScheme = getColorsScheme(editorColorsScheme);
            HighlightInfoType.HighlightInfoTypeImpl infoType = severityRegistrar.getHighlightInfoTypeBySeverity(severity);
            TextAttributesKey key = infoType.getAttributesKey();
            return colorsScheme.getAttributes(key);
        } else {
            return textAttributes;
        }
    }

    @Nullable
    public static Annotation createAnnotation(@NotNull AnnotationHolder holder, @NotNull HighlightSeverity severity, @Nullable TextAttributes forcedTextAttributes, @NotNull TextRange range, @NotNull String message) {
        Annotation annotation;
        if (forcedTextAttributes == null) {
            annotation = severity.equals(HighlightSeverity.ERROR) ? holder.createErrorAnnotation(range, message) : holder.createWarningAnnotation(range, message);
        } else {
            annotation = holder.createAnnotation(severity, range, message);
            annotation.setEnforcedTextAttributes(forcedTextAttributes);
        }
        annotation.setNeedsUpdateOnTyping(false);
        return annotation;
    }
}

File: src/main/kotlin/tlint/InspectionUtil.kt
package tlint

import com.intellij.codeInsight.daemon.impl.SeverityRegistrar
import com.intellij.lang.annotation.Annotation
import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.HighlightSeverity
import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.editor.colors.EditorColorsScheme
import com.intellij.openapi.editor.markup.TextAttributes
import com.intellij.openapi.util.TextRange

object InspectionUtil {

    private fun getColorsScheme(customScheme: EditorColorsScheme?): EditorColorsScheme {
        return customScheme ?: EditorColorsManager.getInstance().globalScheme
    }

    fun getTextAttributes(editorColorsScheme: EditorColorsScheme?, severityRegistrar: SeverityRegistrar, severity: HighlightSeverity): TextAttributes {
        val textAttributes = severityRegistrar.getTextAttributesBySeverity(severity)
        return if (textAttributes == null) {
            val colorsScheme = getColorsScheme(editorColorsScheme)
            val infoType = severityRegistrar.getHighlightInfoTypeBySeverity(severity)
            val key = infoType.attributesKey
            colorsScheme.getAttributes(key)
        } else {
            textAttributes
        }
    }

    fun createAnnotation(holder: AnnotationHolder, severity: HighlightSeverity, forcedTextAttributes: TextAttributes?, range: TextRange, message: String): Annotation? {
        val annotation: Annotation
        if (forcedTextAttributes == null) {
            annotation = if (severity == HighlightSeverity.ERROR) holder.createErrorAnnotation(range, message) else holder.createWarningAnnotation(range, message)
        } else {
            annotation = holder.createAnnotation(severity, range, message)
            annotation.enforcedTextAttributes = forcedTextAttributes
        }
        annotation.setNeedsUpdateOnTyping(false)
        return annotation
    }
}
